docs/modules/ROOT/pages/layer_by_layer.adoc

```diff
@@ -0,0 +1,86 @@
+
+== HAL version
+
+To simplify our application, we can use the HAL instead. The HAL exposes higher level APIs that handle details such
+
+* Automatically enabling the peripheral clock when you're using the peripheral
+* Deriving and applying register configuration from higher level types
+* Implementing the embedded-hal traits to make peripherals useful in third party drivers
+
+The HAL example is shown below:
+
+[source,rust]
+----
+include::example$layer-by-layer/blinky-hal/src/main.rs[]
+----
+
+As you can see, the application becomes a lot simpler, even without using any async code. The `Input` and `Output` hides all the details accessing the GPIO registers, and allow you to use a much simpler API to query the state of the button and toggle the LED output accordingly.
+
+The same downside from the PAC example still applies though: the application is busy looping and consuming more power than necessary.
+
+== Interrupt driven
+
+To save power, we need to configure the application so that it can be notified when the button is pressed using an interrupt.
+
+Once the interrupt is configured, the application can instruct the microcontroller to enter a sleep mode, consuming very little power.
+
+Given Embassy focus on async Rust (which we'll come back to after this example), the example application must use a combination of the HAL and PAC in order to use interrupts. For this reason, the application also contains some helper functions to access the PAC (not shown below).
+
+[source,rust]
+----
+include::example$layer-by-layer/blinky-irq/src/main.rs[lines="1..57"]
+----
+
+The simple application is now more complex again, primarily because of the need to keep the button and LED states in the global scope where it is accessible by the main application loop, as well as the interrupt handler.
+
+To do that, the types must be guarded by a mutex, and interrupts must be disabled whenever we are accessing this global state to gain access to the peripherals.
+
+Luckily, there is an elegant solution to this problem when using Embassy.
+
+== Async version
+
+It's time to use the Embassy capabilities to its fullest. At the core, Embassy has an async excecutor, or a runtime for async tasks if you will. The executor polls a set of tasks (defined at compile time), and whenever a task `blocks`, the executor will run another task, or put the microcontroller to sleep.
+
+[source,rust]
+----
+include::example$layer-by-layer/blinky-async/src/main.rs[]
+----
+
+The async version looks very similar to the HAL version, apart from a few minor details:
+
+* The main entry point is annotated with a different macro and has an async type signature. This macro creates and starts an Embassy runtime instance and launches the main application task. Using the `Spawner` instance, the application may spawn other tasks.
```

Comment:
Do we want to target stable Rust or nightly with this doc? Can’t use the macros with stable, right?

Reply:
Yup, macros require nightly, which I think personally is better because of the reduced boilerplate